### PR TITLE
[berkeley] Use Proof_cache for more transaction snark tests, make tests deterministic

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR builds upon #14031, using the `Proof_cache` mechanism to cache more proofs generated in the transaction snark tests.

In order for this to work, this PR also replaces some uses of the non-deterministic `Private_key.create` with the `Private_key.gen` quickcheck generator, so that we achieve cache hits for all of the relevant tests.

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Comment by mrmr1993** [comment]
!ci-build-me

File: src/lib/transaction_snark/test/zkapp_preconditions/zkapp_preconditions.ml [hunk 1]
```diff
@@ -993,4 +1011,11 @@ let%test_module "Account precondition tests" =
                   failwith
                     "Expected transaction to fail due to invalid account \
                      precondition in the fee payer" ) )
+
+    let () =
+      match Sys.getenv_opt "PROOF_CACHE_OUT" with
```
Comment (nholland94) [inline src/lib/transaction_snark/test/zkapp_preconditions/zkapp_preconditions.ml:1016]: What happens here if we run unit tests with `-only-test` flags? Would we get a proof cache out at the end which is missing some of the necessary proofs?

**Review by nholland94 — APPROVED** [review]

File: src/lib/transaction_snark/test/zkapp_preconditions/zkapp_preconditions.ml (hunk 1 above)
Comment (mrmr1993) [reply src/lib/transaction_snark/test/zkapp_preconditions/zkapp_preconditions.ml:1016]: Only if you also set the `PROOF_CACHE_OUT` environment variable (see [here](https://github.com/MinaProtocol/mina/pull/14031/files#diff-e4944a1c552b2b345674ed2c322f75eac44e78cc0328ada6fe603d5098084f0a)).

Comment (mrmr1993) [reply src/lib/transaction_snark/test/zkapp_preconditions/zkapp_preconditions.ml:1016]: So basically: don't do that :)

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Pull request merged**